1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,445 @@
+explain all the things outside of the message, the stuff peripheral to the
+actual data. In a programming language this would be class names, method names,
+possible exceptions, but would not cover anything about the arguments: type
+hints, possible validations, etc.
+
+The data model is used to describe the contents of the message, which is
+probably what you find in the HTTP body. The word "schema" is often associated
+with the data model too. They are often just two different terms describing the
+same concept, but often the term "schema" is used to represent the actual
+technical file containing the description language, and the term data model is
+more theoretical.
+
+With this in mind, lets go look at how contracts are written for endpoint-based
+APIs (REST, RESTish, some RPC), GraphQL and gRPC.
+
+==== Endpoint-based APIs: OpenAPI & JSON Schema
+
+In the endpoint-based API world there were a few such as
+https://apiblueprint.org/[API Blueprint], https://raml.org/[RAML], and Swagger,
+but for years the tooling was a bit lacking, and mostly only allowed for
+outputting as documentation. Around 2018 one description language settled the
+mainstream favourite, and that was OpenAPI v3.0. Another popular language is
+http://json-schema.org/[JSON Schema], which parts of OpenAPI are based on, and
+mostly compatible with.
+
+----
+The OpenAPI Specification (OAS) defines a standard, programming
+language-agnostic interface description for REST APIs, which allows both humans
+and computers to discover and understand the capabilities of a service without
+requiring access to source code, additional documentation, or inspection of
+network traffic.
+----
+
+OpenAPI (formerly known as Swagger) has become the go-to description language
+for writing "endpoint-based API" contracts, and you will come to see why as it
+pops up in pretty much every part of this book. Planning, design, testing,
+management, you name it, OpenAPI can help out.
+
+.An overly simplified example of OpenAPI describing an API which lists collections and resources of hats.
+[source,yaml]
+----
+openapi: 3.0.2
+info:
+  title: Cat on the Hat API
+  version: 1.0.0
+  description: The API for selling hats with pictures of cats.
+servers:
+  - url: "https://hats.example.com"
+    description: Production server
+  - url: "https://hats-staging.example.com"
+    description: Staging server
+
+paths:
+  /hats:
+    get:
+      description: Returns all hats from the system that the user has access to
+      responses:
+        '200':
+          description: A list of hats.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/hats'
+
+components:
+  schemas:
+    hats:
+      type: array
+      items:
+        $ref: "#/components/schemas/hat"
+
+    hat:
+      type: object
+      properties:
+        id:
+          type: string
+          format: uuid
+        name:
+          type: enum
```

Comment:
```suggestion
          type: string
```

There is no `enum` type. At least swagger editor fails on this line stating:

> Structural error at components.schemas.hat.properties.name.type
should be equal to one of the allowed values
allowedValues: array, boolean, integer, number, object, string

type of this property is `string` but validation rule is made by enum only allowing the listed values.